### ISSUE TITLE
Postgres: Support for multiline string literals causes incorrect parsing behaviour and false warnings from L039

### ISSUE THREAD

**Issue by derickl** [sqlfluff/sqlfluff#2495]
## Expected Behaviour

I expect the following query to lint successfully for the Postgres dialect
```sql
SELECT
    a,
    CASE
        WHEN 1 THEN 'one'
        WHEN 2 THEN 'two'
        ELSE 'other'
    END AS b
FROM test;
```

## Observed Behaviour
```shell
sqlfluff lint -n sample.sql --dialect postgres
== [sample.sql] FAIL                                                                                                                                                         
L:   5 | P:   1 | L039 | Unnecessary whitespace found.
L:   6 | P:   1 | L039 | Unnecessary whitespace found.
L:   7 | P:   1 | L039 | Unnecessary whitespace found.
```
Running `sqlfluff fix -n sample.sql --dialect postgres` yields an output that doesn't pass linting.

From this [thread](https://sqlfluff.slack.com/archives/C01GX4L213J/p1643299089547419), it seems this is an unintended bug from #2356.

The above query lints successfully for the `ansi` dialect.

## Dialect
postgres

## Version
0.9.3
master

## Configuration
Default

**Comment by tunetheweb** [comment]
Here's the output of `sqlfluff parse` for postgres compared with ANSI:

![image](https://user-images.githubusercontent.com/10931297/151403301-38f77dc1-6a28-4428-b01b-6c10a626d64a.png)

You can see Postgres thinks the newline is part of the expression.

This causes it to think the next line is not on a newline, so it doesn't require indenting. Running `sqlfluff fix` then produces the following:

```
SELECT
    a,
    CASE
        WHEN 1 THEN 'one'
 WHEN 2 THEN 'two'
 ELSE 'other'
 END AS b
FROM test;
```

This is due to #2495 which introduced this:

https://github.com/sqlfluff/sqlfluff/blob/7fbc8768a64590c1555c2d91d7bfb3e4741aebea/src/sqlfluff/dialects/dialect_postgres.py#L234-L247

Delimited is greedy and will grab it all. Unfortunately you cannot set `allow_trailing=False` as then it won't allow a newline after, so then it fails to parse.

Think we need to rewrite this not using Delimited so it can specifically exclude the final Newline.

**Comment by johnpkennedy** [comment]
This appears to be a regression bug that was introduced in 0.9.2. It is causing a lot of false positives in Redshift.

**Comment by tunetheweb** [comment]
PR raised - #2526
